blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitReadSaveService.java

```diff
@@ -0,0 +1,203 @@
+        TYPE = type;
+    }
+
+    private GitReadSaveRequest makeSaveRequest(
+            Item item,String branch, String commitMessage,
+            String sourceBranch, String filePath, byte[] contents) {
+        String defaultBranch = "master";
+        GitSCMSource gitSource = null;
+        if (item instanceof MultiBranchProject<?,?>) {
+            MultiBranchProject<?,?> mbp = (MultiBranchProject<?,?>)item;
+            for (SCMSource s : mbp.getSCMSources()) {
+                if (s instanceof GitSCMSource) {
+                    gitSource = (GitSCMSource)s;
+                }
+            }
+        }
+
+        switch(TYPE) {
+            case CLONE:
+                return new GitCloneReadSaveRequest(
+                    gitSource,
+                    StringUtils.defaultIfEmpty(branch, defaultBranch),
+                    commitMessage,
+                    StringUtils.defaultIfEmpty(sourceBranch, defaultBranch),
+                    filePath,
+                    contents
+                );
+            case CACHE_CLONE:
+                return new GitCacheCloneReadSaveRequest(
+                    gitSource,
+                    StringUtils.defaultIfEmpty(branch, defaultBranch),
+                    commitMessage,
+                    StringUtils.defaultIfEmpty(sourceBranch, defaultBranch),
+                    filePath,
+                    contents
+                );
+            default:
+                return new GitBareRepoReadSaveRequest(
+                    gitSource,
+                    StringUtils.defaultIfEmpty(branch, defaultBranch),
+                    commitMessage,
+                    StringUtils.defaultIfEmpty(sourceBranch, defaultBranch),
+                    filePath,
+                    contents
+                );
+        }
+    }
+
+    private GitReadSaveRequest makeSaveRequest(Item item, StaplerRequest req) {
+        return makeSaveRequest(item,
+            req.getParameter("branch"),
+            req.getParameter("commitMessage"),
+            req.getParameter("sourceBranch"),
+            req.getParameter("path"),
+            Base64.decode(req.getParameter("contents"))
+        );
+    }
+
+    private GitReadSaveRequest makeSaveRequest(Item item, JSONObject json) {
+        JSONObject content = json.getJSONObject("content");
+        return makeSaveRequest(item,
+            content.getString("branch"),
+            content.getString("message"),
+            content.getString("sourceBranch"),
+            content.getString("path"),
+            Base64.decode(content.getString("base64Data"))
+        );
+    }
+
+    @Override
+    public Object getContent(@Nonnull StaplerRequest req, @Nonnull Item item) {
+        item.checkPermission(Permission.READ);
+        User user = User.current();
+        if (user == null) {
+            throw new ServiceException.UnauthorizedException("Not authenticated");
+        }
+
+        try {
+            GitReadSaveRequest r = makeSaveRequest(item, req);
```

Comment:
@vivek yes, those aren't fully implemented, I want to have `GitBareRepoReadSaveRequest` be the only option, but it needs to be tested with different people's setups, since it does some wacky stuff with the bare repo; I'll get to them in the next couple days